package jp.co.sss.lms.ct.f02_faq;

import static jp.co.sss.lms.ct.util.WebDriverUtils.*;
import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;
import java.util.List;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.MethodOrderer.OrderAnnotation;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;


/**
 * 結合テスト よくある質問機能
 * ケース05
 * @author holy
 */
@TestMethodOrder(OrderAnnotation.class)
@DisplayName("ケース05 キーワード検索 正常系")
public class Case05 {

	/** 前処理 */
	@BeforeAll
	static void before() {
		createDriver();
	}

	/** 後処理 */
	@AfterAll
	static void after() {
		closeDriver();
	}

	@Test
	@Order(1)
	@DisplayName("テスト01 トップページURLでアクセス")
	void test01() {
		goTo("http://localhost:8080/lms/");
		assertEquals("ログイン | LMS", webDriver.getTitle());
		getEvidence(new Object(){});
	}

	@Test
	@Order(2)
	@DisplayName("テスト02 初回ログイン済みの受講生ユーザーでログイン")
	void test02() {
		WebElement id = webDriver.findElement(By.id("loginId"));
		id.clear();
		id.sendKeys("StudentAA01");
		WebElement password = webDriver.findElement(By.id("password"));
		password.clear();
		password.sendKeys("StudentAA01test");
		WebElement login = webDriver.findElement(By.cssSelector(".btn.btn-primary"));
		login.click();

		assertEquals("コース詳細 | LMS", webDriver.getTitle());
		
		getEvidence(new Object(){});
	}
	
	@Test
	@Order(3)
	@DisplayName("テスト03 上部メニューの「ヘルプ」リンクからヘルプ画面に遷移")
	void test03() {
		WebElement function = webDriver.findElement(By.className("dropdown-toggle"));
		function.click();
		WebElement help = webDriver.findElement(By.linkText("ヘルプ"));
		help.click();
		
		assertEquals("ヘルプ | LMS",webDriver.getTitle());
		
		getEvidence(new Object(){});
	}

	@Test
	@Order(4)
	@DisplayName("テスト04 「よくある質問」リンクからよくある質問画面を別タブに開く")
	void test04() {
		String originalWindow = webDriver.getWindowHandle();
		
		WebElement faq = webDriver.findElement(By.linkText("よくある質問"));
		faq.click();
		
		for (String windowHandle : webDriver.getWindowHandles()) {
		    if (!originalWindow.contentEquals(windowHandle)) {
		    	webDriver.switchTo().window(windowHandle);
		        break;
		    }
		}
		
		assertEquals("よくある質問 | LMS",webDriver.getTitle());
		
		getEvidence(new Object(){});
	}
	
	@Test
	@Order(5)
	@DisplayName("テスト05 キーワード検索で該当キーワードを含む検索結果だけ表示")
	void test05() throws IOException{
		WebElement keyword = webDriver.findElement(By.id("form"));
		keyword.clear();
		keyword.sendKeys("研修");
		//検索ボタンを押下
		WebElement search = webDriver.findElement(By.cssSelector(".btn.btn-primary[value='検索']"));
		search.click();
		
		//findElementsで要素を複数取得する
		final List<WebElement> question = webDriver.findElements(By.className("mb10"));
		String firstText = question.get(0).getText();
		String secondText = question.get(1).getText();
		
		assertEquals("Q.助成金書類の作成方法が分かりません",firstText);
		assertEquals("Q.研修の申し込みはどのようにすれば良いですか？",secondText);
		
		//キャプチャのため500ピクセル下へスクロール
		scrollBy("500");
		getEvidence(new Object(){});
		
		//スクロールをもとに戻す
		scrollBy("-500");
	}
	
	@Test
	@Order(6)
	@DisplayName("テスト06 「クリア」ボタン押下で入力したキーワードを消去")
	void test06() {
		//クリアボタンを押下
		WebElement clear = webDriver.findElement(By.cssSelector(".btn.btn-primary[value='クリア']"));
		clear.click();
		
		WebElement keyword = webDriver.findElement(By.className("form-control"));
		assertEquals("",keyword.getText());
		
		getEvidence(new Object(){});
	}

}
